File: Packages/RallyUI/Sources/RallyUI/Components/SchoolHeader.swift
import SwiftUI
import RallyCore

/// School branding banner displaying the logo, school name, and mascot, themed
/// with the school's primary colors.
///
/// Usage:
/// ```swift
/// SchoolHeader(school: school)
/// ```
public struct SchoolHeader: View {

    // MARK: - Style

    /// Layout variants for the header.
    public enum Style {
        /// Full-width banner with large logo and mascot image.
        case banner
        /// Compact inline display for navigation bars or list headers.
        case compact
    }

    // MARK: - Properties

    @Environment(ThemeEngine.self) private var themeEngine

    private let school: School
    private let style: Style

    // MARK: - Init

    /// Creates a school header.
    /// - Parameters:
    ///   - school: The school to display branding for.
    ///   - style: Layout variant (default `.banner`).
    public init(school: School, style: Style = .banner) {
        self.school = school
        self.style = style
    }

    // MARK: - Body

    public var body: some View {
        switch style {
        case .banner:
            bannerLayout
        case .compact:
            compactLayout
        }
    }

    // MARK: - Banner Layout

    private var bannerLayout: some View {
        ZStack(alignment: .bottomLeading) {
            // Background gradient from school colors
            LinearGradient(
                colors: [
                    themeEngine.activeTheme.primaryColor,
                    themeEngine.activeTheme.secondaryColor
                ],
                startPoint: .topLeading,
                endPoint: .bottomTrailing
            )
            .frame(height: 180)
            .overlay(alignment: .trailing) {
                // Mascot image (decorative, faded)
                mascotImage
                    .opacity(0.2)
                    .frame(maxHeight: 160)
                    .clipped()
            }

            // Logo + school name overlay
            HStack(spacing: RallySpacing.smMd) {
                schoolLogo(size: 56)

                VStack(alignment: .leading, spacing: RallySpacing.xs) {
                    Text(school.name)
                        .font(RallyTypography.sectionHeader)
                        .foregroundStyle(.white)
                        .lineLimit(2)

                    Text(school.mascot)
                        .font(RallyTypography.subtitle)
                        .foregroundStyle(.white.opacity(0.8))
                }
            }
            .padding(RallySpacing.md)
        }
        .clipShape(RoundedRectangle(cornerRadius: RadiusToken.card, style: .continuous))
        .rallyCardShadow()
        .accessibilityElement(children: .ignore)
        .accessibilityLabel("\(school.name) \(school.mascot)")
    }

    // MARK: - Compact Layout

    private var compactLayout: some View {
        HStack(spacing: RallySpacing.sm) {
            schoolLogo(size: 36)

            VStack(alignment: .leading, spacing: 2) {
                Text(school.name)
                    .font(RallyTypography.cardTitle)
                    .foregroundStyle(.white)
                    .lineLimit(1)

                Text(school.mascot)
                    .font(RallyTypography.caption)
                    .foregroundStyle(RallyColors.gray)
                    .lineLimit(1)
            }

            Spacer()

            // School abbreviation badge
            Text(school.abbreviation)
                .font(RallyTypography.caption)
                .fontWeight(.bold)
                .foregroundStyle(themeEngine.activeTheme.primaryColor)
                .padding(.horizontal, RallySpacing.sm)
                .padding(.vertical, RallySpacing.xs)
                .background(
                    Capsule()
                        .fill(themeEngine.activeTheme.primaryColor.opacity(0.15))
                )
        }
        .accessibilityElement(children: .ignore)
        .accessibilityLabel("\(school.name) \(school.mascot)")
    }

    // MARK: - Subviews

    @ViewBuilder
    private func schoolLogo(size: CGFloat) -> some View {
        if let logoURL = school.logoURL {
            AsyncImage(url: logoURL) { phase in
                switch phase {
                case .success(let image):
                    image
                        .resizable()
                        .aspectRatio(contentMode: .fit)
                case .failure:
                    logoPlaceholder(size: size)
                case .empty:
                    ProgressView()
                        .frame(width: size, height: size)
                @unknown default:
                    logoPlaceholder(size: size)
                }
            }
            .frame(width: size, height: size)
            .clipShape(Circle())
        } else {
            logoPlaceholder(size: size)
        }
    }

    private func logoPlaceholder(size: CGFloat) -> some View {
        Circle()
            .fill(themeEngine.activeTheme.primaryColor.opacity(0.3))
            .frame(width: size, height: size)
            .overlay(
                Text(school.abbreviation.prefix(2))
                    .font(size > 40 ? RallyTypography.cardTitle : RallyTypography.caption)
                    .fontWeight(.bold)
                    .foregroundStyle(.white)
            )
    }

    @ViewBuilder
    private var mascotImage: some View {
        if let mascotURL = school.mascotImageURL {
            AsyncImage(url: mascotURL) { phase in
                if let image = phase.image {
                    image
                        .resizable()
                        .aspectRatio(contentMode: .fit)
                }
            }
        }
    }
}

// MARK: - Preview

#Preview("School Header") {
    let school = School(
        id: "demo-school",
        name: "State University",
        mascot: "Wildcats",
        abbreviation: "STU",
        theme: SchoolTheme(
            primaryColor: "#FF6B35",
            secondaryColor: "#131B2E",
            accentColor: "#2D9CDB"
        )
    )

    VStack(spacing: 24) {
        SchoolHeader(school: school, style: .banner)
        SchoolHeader(school: school, style: .compact)
            .padding(.horizontal)
    }
    .padding()
    .background(RallyColors.navy)
    .environment(ThemeEngine())
}

File: Android/ui/src/main/java/com/vanwagner/rally/ui/components/SchoolHeader.kt
package com.vanwagner.rally.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.tooling.preview.Preview
import androidx.compose.ui.unit.dp
import com.vanwagner.rally.ui.theme.RallyColors
import com.vanwagner.rally.ui.theme.RallyRadius
import com.vanwagner.rally.ui.theme.RallySpacing
import com.vanwagner.rally.ui.theme.RallyTypography

// ---------------------------------------------------------------------------
// SchoolHeader
// School branding banner displaying logo, school name, and mascot, themed
// with the school's primary colors.
// Styles: Banner (full-width gradient) and Compact (inline row).
// ---------------------------------------------------------------------------

/** Layout variants for the header. */
enum class SchoolHeaderStyle {
    /** Full-width banner with large logo and gradient background. */
    Banner,
    /** Compact inline display for navigation bars or list headers. */
    Compact,
}

/**
 * School branding header.
 *
 * @param schoolName    The school display name.
 * @param mascot        The school mascot name.
 * @param abbreviation  Short abbreviation (e.g. "STU").
 * @param primaryColor  The school's primary brand color.
 * @param secondaryColor The school's secondary brand color.
 * @param modifier      Optional [Modifier].
 * @param style         Layout variant (default [SchoolHeaderStyle.Banner]).
 * @param logoUrl       Optional URL string for the school logo (placeholder rendered if null).
 */
@Composable
fun SchoolHeader(
    schoolName: String,
    mascot: String,
    abbreviation: String,
    primaryColor: Color,
    secondaryColor: Color,
    modifier: Modifier = Modifier,
    style: SchoolHeaderStyle = SchoolHeaderStyle.Banner,
) {
    when (style) {
        SchoolHeaderStyle.Banner -> SchoolHeaderBanner(
            schoolName = schoolName,
            mascot = mascot,
            abbreviation = abbreviation,
            primaryColor = primaryColor,
            secondaryColor = secondaryColor,
            modifier = modifier,
        )
        SchoolHeaderStyle.Compact -> SchoolHeaderCompact(
            schoolName = schoolName,
            mascot = mascot,
            abbreviation = abbreviation,
            primaryColor = primaryColor,
            modifier = modifier,
        )
    }
}

// -- Banner layout --

@Composable
private fun SchoolHeaderBanner(
    schoolName: String,
    mascot: String,
    abbreviation: String,
    primaryColor: Color,
    secondaryColor: Color,
    modifier: Modifier = Modifier,
) {
    Box(
        modifier = modifier
            .fillMaxWidth()
            .height(180.dp)
            .clip(RoundedCornerShape(RallyRadius.card))
            .background(
                Brush.linearGradient(
                    colors = listOf(primaryColor, secondaryColor),
                )
            )
            .semantics { contentDescription = "$schoolName $mascot" },
        contentAlignment = Alignment.BottomStart,
    ) {
        Row(
            modifier = Modifier
                .padding(RallySpacing.md),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(RallySpacing.smMd),
        ) {
            // Logo placeholder (circle with abbreviation)
            LogoPlaceholder(
                abbreviation = abbreviation,
                primaryColor = primaryColor,
                size = 56,
            )

            Column(verticalArrangement = Arrangement.spacedBy(RallySpacing.xs)) {
                Text(
                    text = schoolName,
                    style = RallyTypography.sectionHeader,
                    color = Color.White,
                    maxLines = 2,
                    overflow = TextOverflow.Ellipsis,
                )
                Text(
                    text = mascot,
                    style = RallyTypography.subtitle,
                    color = Color.White.copy(alpha = 0.8f),
                )
            }
        }
    }
}

// -- Compact layout --

@Composable
private fun SchoolHeaderCompact(
    schoolName: String,
    mascot: String,
    abbreviation: String,
    primaryColor: Color,
    modifier: Modifier = Modifier,
) {
    Row(
        modifier = modifier
            .fillMaxWidth()
            .semantics { contentDescription = "$schoolName $mascot" },
        verticalAlignment = Alignment.CenterVertically,
        horizontalArrangement = Arrangement.spacedBy(RallySpacing.sm),
    ) {
        LogoPlaceholder(
            abbreviation = abbreviation,
            primaryColor = primaryColor,
            size = 36,
        )

        Column(
            modifier = Modifier.weight(1f),
            verticalArrangement = Arrangement.spacedBy(2.dp),
        ) {
            Text(
                text = schoolName,
                style = RallyTypography.cardTitle,
                color = Color.White,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
            )
            Text(
                text = mascot,
                style = RallyTypography.caption,
                color = RallyColors.Gray,
                maxLines = 1,
                overflow = TextOverflow.Ellipsis,
            )
        }

        // Abbreviation badge
        Text(
            text = abbreviation,
            style = RallyTypography.caption.copy(fontWeight = FontWeight.Bold),
            color = primaryColor,
            modifier = Modifier
                .background(
                    color = primaryColor.copy(alpha = 0.15f),
                    shape = RoundedCornerShape(50),
                )
                .padding(horizontal = RallySpacing.sm, vertical = RallySpacing.xs),
        )
    }
}

// -- Shared logo placeholder --

@Composable
private fun LogoPlaceholder(
    abbreviation: String,
    primaryColor: Color,
    size: Int,
) {
    val textStyle = if (size > 40) RallyTypography.cardTitle else RallyTypography.caption

    Box(
        modifier = Modifier
            .size(size.dp)
            .clip(CircleShape)
            .background(primaryColor.copy(alpha = 0.3f)),
        contentAlignment = Alignment.Center,
    ) {
        Text(
            text = abbreviation.take(2),
            style = textStyle.copy(fontWeight = FontWeight.Bold),
            color = Color.White,
        )
    }
}

// MARK: - Previews

@Preview(name = "SchoolHeader - Banner & Compact", showBackground = true, backgroundColor = 0xFF131B2E)
@Composable
private fun SchoolHeaderPreview() {
    Surface(color = RallyColors.Navy) {
        Column(
            modifier = Modifier.padding(RallySpacing.md),
            verticalArrangement = Arrangement.spacedBy(RallySpacing.lg),
        ) {
            SchoolHeader(
                schoolName = "State University",
                mascot = "Wildcats",
                abbreviation = "STU",
                primaryColor = RallyColors.Orange,
                secondaryColor = RallyColors.Navy,
                style = SchoolHeaderStyle.Banner,
            )
            SchoolHeader(
                schoolName = "State University",
                mascot = "Wildcats",
                abbreviation = "STU",
                primaryColor = RallyColors.Orange,
                secondaryColor = RallyColors.Navy,
                style = SchoolHeaderStyle.Compact,
            )
        }
    }
}
